programs/sbf/rust/abiv2/src/lib.rs

```diff
@@ -0,0 +1,121 @@
+use {
+    solana_program::{
+        custom_heap_default, custom_panic_default, entrypoint::SUCCESS, pubkey::Pubkey,
+    },
+    std::{slice, str::FromStr},
+};
+
+const VIRTUAL_ADDRESS_BITS: usize = 32;
+const MM_REGION_SIZE: u64 = 1 << VIRTUAL_ADDRESS_BITS;
+const SCRATCHPAD_ADDR: u64 = MM_REGION_SIZE * 5;
+const CPI_SCRATCHPAD_ADDRESS: u64 = MM_REGION_SIZE * 255;
+
+/// `ReturnDataScratchPad` contains the return data. Programs will be allowed to read and write to
+/// the data slice.
+#[repr(C)]
+struct ReturnDataScratchPad<'a> {
+    /// Pubkey of the last program to write to the scratchpad slice.
+    pubkey: Pubkey,
+    data: &'a [u8],
+}
+
+/// `TransactionAccount` is the generic representation for an account in the transaction.
+#[repr(C)]
+struct TransactionAccount<'a> {
+    pubkey: Pubkey,
+    owner: Pubkey,
+    lamports: u64,
+    payload: &'a [u8],
+}
+
+/// `ExecutionData` is split from `TransactionContext` to facilitate reading from bytes.
+#[repr(C)]
+struct ExecutionData<'a> {
+    return_data_scratchpad: ReturnDataScratchPad<'a>,
+    cpi_scratchpad: &'a [u8],
+    /// Index in transaction of the currently executing instruction
+    ix_idx: u64,
+    /// Number of instructions in transaction
+    ix_num: u64,
+}
+
+/// `TransactionContext` contains the information extracted from the transaction area
+#[repr(C)]
+struct TransactionContext<'a> {
+    execution_data: &'a ExecutionData<'a>,
+    accounts: &'a [TransactionAccount<'a>],
+}
+
+impl TransactionContext<'_> {
+    fn new(mut addr: *const u8) -> TransactionContext<'static> {
```

Comment:
This is a glimpse of what a future entry point should look like cc @febo and @joncinque